crates/iceberg/src/writer/base_writer/data_file_writer.rs

```diff
@@ -0,0 +1,323 @@
+                    .with_metadata(HashMap::from([(
+                        PARQUET_FIELD_ID_META_KEY.to_string(),
+                        "5".to_string(),
+                    )])),
+            ]
+            .into(),
+            vec![Arc::new(Int64Array::from_iter_values(vec![1; 1024]))],
+            None,
+        ));
+        let col2 = Arc::new(arrow_array::StringArray::from_iter_values(vec![
+            "test";
+            1024
+        ])) as ArrayRef;
+        let col3 = Arc::new({
+            let list_parts = arrow_array::ListArray::from_iter_primitive::<Int64Type, _, _>(vec![
+                Some(
+                    vec![Some(1),]
+                );
+                1024
+            ])
+            .into_parts();
+            arrow_array::ListArray::new(
+                Arc::new(list_parts.0.as_ref().clone().with_metadata(HashMap::from([(
+                    PARQUET_FIELD_ID_META_KEY.to_string(),
+                    "6".to_string(),
+                )]))),
+                list_parts.1,
+                list_parts.2,
+                list_parts.3,
+            )
+        }) as ArrayRef;
+        let col4 = Arc::new(StructArray::new(
+            vec![arrow_schema::Field::new(
+                "sub_col",
+                arrow_schema::DataType::Struct(
+                    vec![arrow_schema::Field::new(
+                        "sub_sub_col",
+                        arrow_schema::DataType::Int64,
+                        true,
+                    )
+                    .with_metadata(HashMap::from([(
+                        PARQUET_FIELD_ID_META_KEY.to_string(),
+                        "7".to_string(),
+                    )]))]
+                    .into(),
+                ),
+                true,
+            )
+            .with_metadata(HashMap::from([(
+                PARQUET_FIELD_ID_META_KEY.to_string(),
+                "8".to_string(),
+            )]))]
+            .into(),
+            vec![Arc::new(StructArray::new(
+                vec![
+                    arrow_schema::Field::new("sub_sub_col", arrow_schema::DataType::Int64, true)
+                        .with_metadata(HashMap::from([(
+                            PARQUET_FIELD_ID_META_KEY.to_string(),
+                            "7".to_string(),
+                        )])),
+                ]
+                .into(),
+                vec![Arc::new(Int64Array::from_iter_values(vec![1; 1024]))],
+                None,
+            ))],
+            None,
+        ));
+        let to_write =
+            RecordBatch::try_new(schema.clone(), vec![col0, col1, col2, col3, col4]).unwrap();
+
+        // prepare writer
+        let pb = ParquetWriterBuilder::new(
+            WriterProperties::builder().build(),
+            to_write.schema(),
+            file_io.clone(),
+            location_gen,
+            file_name_gen,
+        );
+        let mut data_file_writer = DataFileWriterBuilder::new(pb)
```

Comment:
It feels a little bit odd to pass a builder to another builder. Could we do something like this?

```rust
DataFileWriter
    .builder()
    // this would call ParquetWriter::builder() to get a `ParquetWriterBuilder`,
    // and then pass ownership of the `DataFileWriter` to the `ParquetWriterBuilder`,
    // returning the `ParquetWriterBuilder`
    .with_writer(ParquetWriter)  
    // these calls happen in the `ParquetWriterBuilder`, 
    // allowing customization of the wrapped concrete writer
    .with_foo()
    .with_bar()
    // this finalizes the `ParquetWriterBuilder`, building a
    // `ParquetWriter`, and returns the `DataFileWriterBuilder`
    // that was passed earlier, after first passing in the `ParrquetWriter`
    .build_writer()
    // these calls now happen on the `DataFileWriterBuilder`,
    // allowing further setup of the `DataFileWriter`
    .with_baz()
    .with_quux()
    // finally returns a `DataFileWriter`, or perhaps a
    // `Result<DataFileWriter>` or Future of one of those
    .build()

Reply:
This top-down way looks good to me to avoid passing builders to each other.  However, I am not sure whether this style will incur more complexity in the future. 

Actually, this API for now can do this thing because it doesn't restrict how to create and use the builder like the following. So I think this is more relative to the implementation style of our builder rather than the API design.
```
#[derive(Clone)]
struct B;

impl B {
    pub fn new() -> Self {
        Self
    }

    pub fn with_config(&mut self, _: ()) -> &mut Self {
        self
    }
}

#[async_trait::async_trait]
impl IcebergWriterBuilder for B {
    type R = BW;
    type C = ();
    async fn build(self, _: Self::C) -> Result<Self::R> {
        Ok(BW)
    }
}

struct BW;

impl BW {
    pub fn builder() -> B {
        B::new()
    }
}

#[async_trait::async_trait]
impl IcebergWriter for BW {
    async fn write(&mut self, _input: DefaultInput) -> Result<()> {
        Ok(())
    }

    async fn flush(&mut self) -> Result<DefaultOutput> {
        Ok(vec![])
    }
}

#[derive(Clone)]
struct A<I> {
    inner: Option<I>,
}

impl<I: IcebergWriterBuilder> A<I> {
    pub fn new() -> Self {
        Self { inner: None }
    }

    pub fn with_buidler(&mut self, builder: I) -> &mut I {
        self.inner = Some(builder);
        self.inner.as_mut().unwrap()
    }

    pub fn with_config(&mut self, _: ()) -> &mut Self {
        self
    }
}

struct AW;

impl AW {
    pub fn builder<I:IcebergWriterBuilder>() -> A<I> {
        A::<I>::new()
    }
}

#[async_trait::async_trait]
impl<I: IcebergWriterBuilder> IcebergWriterBuilder for A<I> {
    type R = AW;
    type C = ();
    async fn build(self, _: Self::C) -> Result<Self::R> {
        Ok(AW)
    }
}

#[async_trait::async_trait]
impl IcebergWriter for AW {
    async fn write(&mut self, _input: DefaultInput) -> Result<()> {
        Ok(())
    }

    async fn flush(&mut self) -> Result<DefaultOutput> {
        Ok(vec![])
    }
}

async fn test() {
    let mut a = AW::builder();
    a.
        // config first A
        with_config(()).
        with_buidler(AW::builder()).
        // config second A
        with_config(()).
        with_buidler(BW::builder()).
        // config BW
        with_config(());
    let writer = a.build(()).await;
}
```

Reply:
Both way looks good to me. cc @Fokko @Xuanwo @liurenjie1024

Reply:
I remembered that we have discussed about this before...

Reply:
Do you mean this https://github.com/apache/iceberg-rust/pull/135#issuecomment-1905325893. Indeed they look similar.🥵

Reply:
Maybe we can move on first if both way looks good?

Reply:
LGTM.